### PR TITLE
fix(be): enforce DKIM tag checks on the DNSSEC path

### PR DESCRIPTION
**Pull request by aterga**
## Problem

A cross-validation audit against the [email-recovery design doc](https://github.com/dfinity/internet-identity/pull/3836) surfaced that the DNSSEC-verified DKIM path bypassed four tag checks that the DoH path enforces. Design §5.4 requires both paths to share the same DKIM tag contract, but the DNSSEC path's `prepare_partial_verification` (signature-only checks at email-arrival time) and `submit_dkim_leaf::run_submit` (DNS-record-aware checks after the DKIM TXT is verified) ran a stripped-down subset.

Concretely, on the DNSSEC path:

- A signature with stale `x=` would verify — the replay window stretches from the DKIM signer's chosen deadline (typically a few hours) to the RRSIG validity (days–weeks).
- A future-dated `t=` would verify — design §5.4 calls this out as a must-enforce.
- A signer with DNS `t=s` and forged `i=alice@evil.com` against `d=victim.com` would verify — AUID was never checked.
- A signature under a key the signer marked `t=y` (testing-mode) would verify against production traffic.

## Changes

Extract the five tag checks into `dkim::tag_checks` as a single source of truth, then call them from both paths:

- **`dkim::tag_checks`** (new module): `check_signature_not_expired`, `check_signature_not_from_future`, `check_subject_signed`, `check_auid_aligned`, `check_dns_not_testing`. Split by *what data they need*, not *which path calls them*: the first three are signature-header-only (callable as soon as `parse_dkim_signature` returns); the last two need the parsed DNS record's flags.
- **DoH path** (`dkim::verify::verify_dkim`): refactored to delegate to the helpers, preserving the existing `DkimCheck` diagnostic trail. No behavioural change.
- **DNSSEC prepare** (`email_recovery::smtp::prepare_partial_verification`): adds the signature-header-only checks at email-arrival time. Reordered the local Subject-in-`h=` check to run *before* body-hash (matching the DoH path's check order, and surfacing tag failures faster).
- **DNSSEC submit** (`email_recovery::submit_leaf::run_submit`): adds the DNS-record-dependent checks right after `parse_dkim_txt`, before the cryptographic signature check. Extracted as `enforce_dns_record_tag_checks` so the wiring is unit-testable without the full DNSSEC chain validator.
- `PartialVerification` and the submit-side `Snapshot` carry `signing_auid` so the submit-time AUID check has the parsed `i=` value it needs.

## Tests

- 12 unit tests in `dkim::tag_checks::tests` cover each helper's discriminating cases (skew window, exact vs subdomain AUID, label-anchored evil-suffix rejection, etc.).
- 4 prepare-path tests in `email_recovery::smtp::tests` verify the wiring at email-arrival time (rejects future-dated `t=`, rejects expired `x=`, rejects missing Subject, accepts `t=` within skew).
- 5 submit-path tests in `email_recovery::submit_leaf::tests` cover `enforce_dns_record_tag_checks` (rejects `t=y`, rejects subdomain AUID under `t=s`, accepts under loose-mode, accepts exact match, rejects evil-suffix).
- The retired duplicate `auid_aligns` tests in `dkim::verify::tests` are subsumed by the new helper-side coverage.

### CODE REVIEW

File: src/internet_identity/src/dkim/verify.rs [hunk 1]
```diff
@@ -177,63 +174,51 @@ fn try_verify_signature(
...
+    if let Err(reason) = check_signature_not_from_future(&sig, now_secs) {
+        let detail = sig
+            .t
+            .map(|t| format!("signature claims t={t}, now={now_secs}"));
+        checks.push(check(
+            DkimCheckName::SignatureNotFromFuture,
+            DkimCheckStatus::Fail,
+            detail,
+        ));
+        return Err((reason, checks));
```
[the remark this answers is not in the document]
Comment (aterga) [reply src/internet_identity/src/dkim/verify.rs:207]: Good catch — restored in 34da7103. Re-imported `CLOCK_SKEW_SECS` from the shared `tag_checks` module (still its single source of truth) and used it in the format string, so the diagnostic again reads `signature claims t=…, now=…, beyond 60s skew`.

File: src/internet_identity/src/email_recovery/submit_leaf.rs [hunk 2]
```diff
@@ -380,6 +393,25 @@ fn run_submit(
     Ok(())
 }
 
+/// Apply the DKIM tag checks that depend on flags published in the
+/// DKIM TXT record. Runs at submit time on the DNSSEC path — the
+/// signature-header-only checks (`x=`, `t=`, Subject in `h=`) have
+/// already run at `smtp::prepare_partial_verification`. Shared with
+/// the DoH path via [`crate::dkim::tag_checks`]; extracted so the
+/// DNSSEC-path wiring is unit-testable without spinning up the full
+/// DNSSEC chain validator.
```
[the remark this answers is not in the document]
Comment (aterga) [reply src/internet_identity/src/email_recovery/submit_leaf.rs]: Right — `tag_checks` is `mod tag_checks` (private), so that intra-doc link doesn't resolve from outside `dkim`. Fixed in 34da7103: pointed at the re-exported helpers (`crate::dkim::check_dns_not_testing` / `check_auid_aligned`) instead.

File: src/internet_identity/src/email_recovery/smtp.rs [hunk 3]
```diff
@@ -441,6 +442,29 @@ fn prepare_partial_verification(
...
+    // Signature-header-only tag enforcement (design §5.4). These are
+    // the same checks the DoH path runs inside `crate::dkim::verify`;
+    // running them here closes the DNSSEC-path parity gap surfaced by
+    // the audit: without these, a sender on a DNSSEC-signed domain
+    // could verify with a stale `x=` (replay window stretches to the
+    // RRSIG validity, days–weeks), a future-dated `t=`, or an `h=`
+    // that doesn't cover `Subject` (nonce-rewrite by a MITM). The DNS
+    // -record-dependent tag checks (`i=` alignment under `t=s`, and
+    // `t=y` testing-mode) run later in `submit_dkim_leaf::run_submit`
+    // once the DKIM leaf has been DNSSEC-verified.
```
[the remark this answers is not in the document]
Comment (aterga) [reply src/internet_identity/src/email_recovery/smtp.rs]: Fixed in 34da7103 — rewrapped so the hyphen no longer sits at the start of its own line.

**Comment by aterga** [comment]
**PR Review: #3911 – fix(be): enforce DKIM tag checks on the DNSSEC path**

**Overall verdict:** Excellent, pragmatic fix. This PR closes a real security gap (identified via cross-validation against the email-recovery design) without unnecessary scope creep. The centralization of the five tag checks into `dkim::tag_checks` is the right architectural move and eliminates the previous divergence between the DoH and DNSSEC paths. Code is clean, tests are comprehensive (12 unit + 9 integration-style), diagnostics are preserved, and the split between header-only vs. DNS-dependent checks is exactly what the two-stage DNSSEC flow requires. No blockers. Auto-merge is appropriate.

**Security assessment**  
- **Gap closed cleanly**: The DNSSEC path previously allowed replay attacks (stale `x=` up to RRSIG lifetime), future-dated signatures (`t=`), AUID forgery under `t=s` strict mode, and production use of `t=y` testing keys. All four (plus the `Subject` hardening) are now enforced uniformly. No new gaps introduced.  
- **No major security issues**: No new dependencies, no unsafe code, no relaxed parsing, no clock skew abuse, no missing error paths. The `now_secs` usage in prepare vs. submit is handled correctly (header-only checks run at arrival time). Copilot security scan came back clean, and the changes are minimal and auditable.  
- **Threat model alignment**: Matches design §5.4 exactly (expiration, future-dating, `Subject` in `h=`, AUID alignment per the crate’s existing RFC 6376 §3.5 logic, and `t=y` rejection).

**Blockers**  
None.

**Easy-to-change improvements (all <30 min)**  
1. **Extract `CLOCK_SKEW_SECS` to a single source of truth**  
   It is already duplicated with the DNSSEC verifier (explicitly noted in the module comment). Move it to `dkim/mod.rs` (or a new `dkim/constants.rs`) as `pub(crate) const CLOCK_SKEW_SECS: u64 = 60;` and re-export/use it from `tag_checks`.  
   *Action*: One-line move + update the two references and the comment. Prevents future skew-policy drift.

2. **Add a one-line regression test for path parity**  
   In `dkim::verify::tests` (or a new test in `email_recovery`), assert that a synthetic signature that fails any of the five checks produces the *exact same* `VerificationFailReason` (and same `DkimCheck` diagnostic) on both the DoH path and the DNSSEC prepare+submit path.  
   *Action*: Add 1 test function using the existing `sig_with` / `dns` helpers. Catches any future refactoring drift with zero runtime cost.

3. **Polish test names and add one missing edge case**  
   The 12 unit tests are solid but some names are generic. Rename a couple for self-documentation (e.g. `auid_relaxed_subdomain_prefix_attack_rejected`). Also add one explicit test for `i=` with no `@` (should treat whole string as domain per current `auid_aligns` logic).  
   *Action*: Quick rename + one new `#[test]` (already covered implicitly, but explicit is better for future readers).

4. **Tiny doc/link fix (already partially addressed in 34da710)**  
   The follow-up commit fixed the typo and restored skew info in the diagnostic. Just double-check that all intra-doc links in `tag_checks.rs` resolve (they point to private submodules).  
   *Action*: `cargo doc --open` and verify (or `grep` for broken links). Trivial.

**Strategic recommendations (longer-term value)**  
1. **Prevent this class of drift forever**  
   The root cause was two separate pipelines without a shared contract. Consider evolving this into a single `dkim::verify::verify_signature` facade that both paths call (passing either a full `DkimDnsRecord` or a partial view). The current split is correct for performance reasons, but a facade would make the “single source of truth” even more obvious to future auditors.

2. **Add property-based testing for the tag contract**  
   One `proptest!` that generates random `DkimSignature` + `DkimDnsRecord` combinations and asserts both paths reject/fail identically would be high ROI. (The current unit tests are excellent but manual.)

3. **Telemetry / observability**  
   After merge, add a quick counter for each new `VerificationFailReason` variant (`SignatureExpired`, `SignatureFutureDated`, `SubjectNotSigned`, `AuidMisaligned`, `TestingMode`). This will immediately surface if any legitimate senders are hitting the newly-enforced rules on the DNSSEC path.

4. **Human review cadence for recovery flows**  
   AI co-author + Copilot review worked great here, but email-recovery code touches the core identity-binding path. For any future changes to this module (or the next recovery-related PR), require 2+ engineer sign-offs + a quick security review. No issue with *this* PR — just a general process recommendation.

**Summary**  
This is the kind of focused, high-signal security PR we want: small surface, big risk reduction, zero regression risk. All feedback above is optional polish. Ship it.

(If you want me to draft any of the suggested test code or the const extraction diff, just say the word.)

**Review by sea-snake — APPROVED** [review]

**Comment by aterga** [comment]
Thanks for the detailed review. Picked the two highest-ROI items:

**#2 — Path-parity regression test** ([8494095f](https://github.com/dfinity/internet-identity/pull/3911/commits/8494095f)). Added `doh_and_dnssec_paths_reject_future_dated_t_with_same_reason` to `email_recovery::smtp::tests`. It runs the same forged `t=now+1000s` signature through both `dkim::verify_dkim` (DoH) and `prepare_partial_verification` (DNSSEC) and asserts both surface `SignatureFutureDated`. Picked future-dated `t=` as the representative case because it short-circuits early in both pipelines without needing a fully-valid signature behind it — keeps the test focused on the parity invariant, not on crypto setup. The five existing DNSSEC-side tests already guard each helper-call wiring independently; this one captures the cross-path property explicitly.

**#4 — Doc-link verification.** Ran `cargo doc -p internet_identity --bin internet_identity --no-deps` and confirmed none of the broken intra-doc warnings come from `tag_checks.rs` or my edits in [34da7103](https://github.com/dfinity/internet-identity/pull/3911/commits/34da7103). The two new helper links (`crate::dkim::check_dns_not_testing` / `check_auid_aligned`) resolve correctly. The remaining warnings in the codebase (`dkim::types.rs:44`, `dmarc::parse.rs`, `doh::cache.rs`, etc.) are pre-existing references to private items and out of scope here.

**Skipped (with reasoning):**

- **#1 — Extract `CLOCK_SKEW_SECS` to `dkim/mod.rs`.** There are two `CLOCK_SKEW_SECS` constants in the tree (`dkim::tag_checks` for DKIM `t=` skew; `dnssec::verify` for RRSIG validity skew). They happen to share the value 60 today but represent semantically distinct concepts — coupling them via a shared constant would create false sharing and obscure the right tuning surface if one ever needs adjustment independently. Within `dkim`, moving the constant from `tag_checks.rs` to `mod.rs` is purely organisational; I'd rather keep it co-located with its only callers.
- **#3 — Test renames + no-`@` AUID case.** The `i=` with no `@` case is already covered by `auid_aligns_handles_no_at_sign` in `tag_checks.rs`. Renames (e.g. `auid_evil_suffix_is_blocked` → `auid_relaxed_subdomain_prefix_attack_rejected`) are subjective — I'd rather not churn the test names mid-PR if the existing ones are reasonably self-documenting.

**Strategic items (1–4):** Worth follow-ups but each is its own PR — a `verify_signature` facade is a meaningful refactor, `proptest` would pull a new dep, telemetry counters need a metrics-design pass, and the review-cadence point is process not code. Happy to spin up any of these as separate work if it's the right time.
